polaris-service/src/test/java/org/apache/polaris/service/catalog/io/TestFileIO.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.service.catalog.io;
+
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.ManifestFile;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+
+/**
+ * File IO wrapper used for tests. It measures the number of bytes read, files written, and files
+ * deleted. It can inject exceptions during InputFile and OutputFile creation.
+ */
+public class TestFileIO implements FileIO {
+  private final FileIO io;
+  private final Supplier<RuntimeException> newInputFileExceptionSupplier;
+  private final Supplier<RuntimeException> newOutputFileExceptionSupplier;
```

Comment:
Are these optional?

Reply:
Marked them as NotNull.

I think the two options are either to allow the supplier to be null or return a null RuntimeException. For now I've gone with the latter so the supplier itself has to be specified.

Reply:
Relying on a supplier that returns null is a bit unintuitive, I think just having the supplier not be required to be present makes the intent more obvious

Reply:
Updated to `Optional<Supplier<RuntimeException>>`